src/trace/context/ContextManager.ts

```diff
@@ -18,19 +18,74 @@
  */
 
 import Context from '../../trace/context/Context';
+import Span from '../../trace/span/Span';
 import SpanContext from '../../trace/context/SpanContext';
 import { AsyncLocalStorage } from 'async_hooks';
 
-const store = new AsyncLocalStorage<Context>();
+type AsyncState = {context: Context, spans: Span[]};
+
+const store = new AsyncLocalStorage<AsyncState>();
 
 class ContextManager {
-  get current(): Context {
-    return store.getStore() || new SpanContext();
+  get asyncState(): AsyncState {
+    let asyncState = store.getStore();
+
+    if (asyncState === undefined) {
+      asyncState = {context: new SpanContext(), spans: []};
+      store.enterWith(asyncState);
+    }
+
+    return asyncState;
+  }
+
+  get current(): Context { return this.asyncState.context; }
+  get spans(): Span[] { return this.asyncState.spans; }
+
+  spans_dup(): Span[] {
+    let asyncState = store.getStore();
+
+    if (asyncState === undefined) {
+      asyncState = {context: new SpanContext(), spans: []};
+    } else {
+      asyncState = {context: asyncState.context, spans: [...asyncState.spans]};
+    }
+
+    store.enterWith(asyncState);
+
+    return asyncState.spans;
   }
 
-  withContext(callback: (...args: any[]) => void, ...args: any[]) {
-    return store.run(new SpanContext(), callback);
+  clear(): void {
+    store.enterWith(undefined as unknown as AsyncState);
   }
+
+  restore(context: Context, spans: Span[]): void {
+    store.enterWith({context, spans: spans || []});
+  }
+
+  withSpan(stopOnSuccess: boolean, span: Span, callback: (...args: any[]) => any, ...args: any[]): any {
+    try {
+      return callback(span, ...args);
+
+    } catch (e) {
+      span.error(e);
+
+      if (!stopOnSuccess) {
+        span.stop();
+      }
+
+      throw e;
+
+    } finally {
+      if (stopOnSuccess) {
+        span.stop();
+      }
+    }
+  }
+
+  // withContext(callback: (...args: any[]) => void, ...args: any[]) {
+  //   return store.run({context: new SpanContext(), spans: []}, callback);
+  // }
```

Comment:
Maybe we can get rid of this `withSpan` now to eliminate the "callbacks", user codes become

```typescript

        const span = ContextManager.current.newExitSpan(operation, host).start();

        span.component = Component.HTTP;
        span.layer = SpanLayer.HTTP;
        span.tag(Tag.httpURL(host + pathname));

        const request = original.apply(this, arguments);

        span.extract().items.forEach((item) => {
          request.setHeader(item.key, item.value);
        });
```

instead of 

```typescript

        const [span, request]: [Span, ClientRequest] = ContextManager.withSpan(false,
            ContextManager.current.newExitSpan(operation, host).start(), (_span: Span, args) => {

          _span.component = Component.HTTP;
          _span.layer = SpanLayer.HTTP;
          _span.tag(Tag.httpURL(host + pathname));

          const _request = original.apply(this, args);

          _span.extract().items.forEach((item) => {
            _request.setHeader(item.key, item.value);
          });

          return [_span, _request];

        }, arguments);
```

I'm sorry it may be `withContext` that misled you to follow the same style 😂

Reply:
Actually that is what I did at first but realized this wrapper could be very handy especially for users to avoid errors. Yes this is a callback but it is a cheap synchronous callback (just a function call), not the more expensive needs-its-own-async-context type of callback. Have a look at how I am using it in the examples I send up and if you do not think it is a net positive then I will remove.